cli_composer.go

```diff
@@ -102,7 +102,78 @@ func composeMessageHeader(replyMsg *fbb.Message) *fbb.Message {
 	return msg
 }
 
-func composeMessage(replyMsg *fbb.Message) {
+func composeMessage(args []string) {
+	set := pflag.NewFlagSet("compose", pflag.ExitOnError)
+	subject := set.StringP("subject", "s", "", "")
+	attachments := set.StringArrayP("attachment", "a", nil, "")
+	ccs := set.StringArrayP("cc", "c", nil, "")
+	set.Parse(args)
+	recipients := set.Args() // Call after Parse()
+
+	// Check if any args are set. If so, go non-interactive
+	// Otherwise, interactive
+	if (len(*subject) + len(*attachments) + len(*ccs) + len(recipients)) > 0 {
+		noninteractiveComposeMessage(*subject, *attachments, *ccs, recipients)
+	} else {
+		interactiveComposeMessage(nil)
+	}
+
+}
+
+func noninteractiveComposeMessage(subject string, attachments []string,
+	ccs []string, recipients []string) {
+	// We have to verify the args here. Follow the same pattern as main()
+	// We'll allow a missing recipient if CC is present (or vice versa)
+	if len(recipients)+len(ccs) <= 0 {
+		fmt.Fprint(os.Stderr, "ERROR: Missing recipients in non-interactive mode!\n")
+		os.Exit(1)
+	}
+
+	// Subject is optional. Print a mailx style warning
+	if subject == "" {
+		fmt.Fprint(os.Stderr, "Warning: missing subject; hope that's OK\n")
+	}
+
+	msg := fbb.NewMessage(fbb.Private, fOptions.MyCall)
+	msg.SetFrom(fOptions.MyCall)
+	for _, to := range recipients {
+		msg.AddTo(to)
+	}
+	for _, cc := range ccs {
+		msg.AddCc(cc)
+	}
+
+	msg.SetSubject(subject)
+
+	// Handle Attachments. Since we're not interactive, treat errors as fatal so the user can fix
+	for _, filename := range attachments {
+		file, err := readAttachment(filename)
+		if err != nil {
+			fmt.Fprint(os.Stderr, err.Error()+"\nAborting! (Message not posted)\n")
+			os.Exit(1)
+		}
+		msg.AddFile(file)
+	}
+
+	// Read the message body from stdin
+	body, _ := ioutil.ReadAll(os.Stdin)
+	if string(body) == "" {
```

Comment:
Maybe use `if len(body) == 0 {` instead.

Reply:
I was wondering about that! I even googled for it! Landed at https://stackoverflow.com/questions/18594330/what-is-the-best-way-to-test-for-an-empty-string-in-go which shrugged it off and said both are idiomatic. Happy to change if you prefer.

Reply:
body is of type `[]byte`, not `string`. So that's why I would prefer `len(body) == 0`, instead of casting it to a string just to check if it's length is zero 🙂 

But no big deal! The Go compiler is smart enough not to do an extra allocation such simple `[]byte` to `string` casts.

Reply:
Ahhhh, I missed the type difference. Thanks for being explicit. PR sent.